File: backend/doggy_dog_diary/schemas/dog.py
from __future__ import annotations

from datetime import date, datetime
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class DogSex(str, Enum):
    male = "male"
    female = "female"
    unknown = "unknown"


class DogStatus(str, Enum):
    current = "current"
    deceased = "deceased"
    rehomed = "rehomed"


class DogBase(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    date_of_birth: date | None = None
    sex: DogSex = DogSex.unknown
    breed: str | None = None
    neutered: bool | None = None
    microchip: str | None = None
    status: DogStatus = DogStatus.current
    status_date: date | None = None
    kc_registered_name: str | None = None
    kc_number: str | None = None
    kc_body: str | None = None
    description: str | None = None


class DogCreate(DogBase):
    pass


class DogUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    date_of_birth: date | None = None
    sex: DogSex | None = None
    breed: str | None = None
    neutered: bool | None = None
    microchip: str | None = None
    status: DogStatus | None = None
    status_date: date | None = None
    kc_registered_name: str | None = None
    kc_number: str | None = None
    kc_body: str | None = None
    description: str | None = None


class DogRead(DogBase):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    profile_photo_path: str | None = None
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/dog.ts
export type DogSex = "male" | "female" | "unknown";
export type DogStatus = "current" | "deceased" | "rehomed";

export interface Dog {
  id: string;
  name: string;
  date_of_birth: string | null;
  sex: DogSex;
  breed: string | null;
  neutered: boolean | null;
  microchip: string | null;
  status: DogStatus;
  status_date: string | null;
  kc_registered_name: string | null;
  kc_number: string | null;
  kc_body: string | null;
  description: string | null;
  profile_photo_path: string | null;
  created_at: string;
  updated_at: string;
}

export type DogCreatePayload = Pick<Dog, "name"> &
  Partial<
    Omit<Dog, "id" | "name" | "profile_photo_path" | "created_at" | "updated_at">
  >;

export type DogUpdatePayload = Partial<
  Omit<Dog, "id" | "profile_photo_path" | "created_at" | "updated_at">
>;

export const DOG_SEX_LABELS: Record<DogSex, string> = {
  male: "Male",
  female: "Female",
  unknown: "Unknown",
};

export const DOG_STATUS_LABELS: Record<DogStatus, string> = {
  current: "Current",
  deceased: "Deceased",
  rehomed: "Rehomed",
};
